### ISSUE TITLE
Feature request: any way to tell from accept handlers whether the update is direct or bubbled?

### ISSUE THREAD

**Issue by rixo** [PepsRyuu/nollup#79]
Use case: Svelte components can include CSS.

~~~html
<script>
  import { name } from './foo.js'
</script>

<p> I am {name}.<p>

<style>
  p { color: green; }
</style>
~~~

When people are fine tuning their CSS, they're likely to trigger a lot of HMR updates and, if I can tell whether this is just a CSS change, I can skip replacing the whole component and get away with just switching a stylesheet in the browser. This is far better because replacing a component always comes with a risk of HMR quirk, while switching a stylesheet is essentially bulletproof...

But I can only do that safely if I'm sure that the HMR update comes from a change to the component file itself. In the example above, if the component's `accept` handler is run because it has bubbled from `./foo.js`, I do have to recreate the component to reflect the change.

Problem: there's no way currently to tell if an accept handler is run because its module has changed or if it is because an HMR update has bubbled from a dependency, is there?

Suggested solution: in my own hot plugin, I'm passing an argument to the accept handlers with the info.

~~~js
import.meta.hot.accept(({ hasChangedDeps }) => {
  if (!hasChangedDeps && onlyCss) onlyInjectCss()
  ...
})
~~~

I absolutely don't recommend the name `hasChangedDeps` because each time I go back to this code it takes me some time to figure out what it means already. Maybe `bubbled` would be good? But the general strategy works well.

That being said, any way to tell if the update is direct or bubbled from an accept handler would be ok for my use case.

**Comment by PepsRyuu** [comment]
Normally this would be achieved by having a ```module.hot.accept``` handler inside the generated CSS module itself: https://github.com/PepsRyuu/rollup-plugin-hot-css/blob/master/index.js#L6 That's usually how all bundlers handle such a situation. Looking through ```svelte-hot``` though, the CSS is inlined which means this is not possible.

There's nothing really difficult about this from a technical perspective. Could easily pass a list of all of the module ids that have been invalidated up until the accept handler. 

```
module.hot.accept(e => {
    if (e.changed.length === 1 && e.changed[0] === module.id) {
        // css only?
    }
});
```

My concern here is what other bundlers are doing, because as far as I'm aware, they don't have such a feature? While we could pass an argument without breaking anything, I'm worried about future changes to these bundlers that could make this incompatible. Has this been discussed for other bundlers at all?

**Comment by rixo** [comment]
No, others they don't have it as far as I can tell. I haven't dug too much, and I haven't asked. I don't have the resources to chase all the prices unfortunately, so I'm focusing on the one I actually want to use first. Like I said, Nollup being the fastest and aiming for full Rollup compatibility, I believe it can be the best solution in the Svelte ecosystem :)

I understand your concern about compatibility with other bundlers, even if I don't particularly share it. Everybody and their cousin are rolling their own HMR API these days... It's about taking the lead here ;)

Anyway, if this is your main concern, how would you feel about adding another `accept` method, with another name, to minimize risks of future conflicts, like `acceptExtra` or anything. I'd be happy with that.

That makes me think of another idea I wanted to float though. How would you feel about making the whole HMR runtime pluggable somehow? I mean something like providing my own `hmr-plugin.js` as an option? This would allow me to implement all my crazy ideas freely without compromising Nollup's prudent stance. Need to dig a little bit into the details, but I'm quite confident this is perfectly feasible, and it would solve the present issue for me. And probably quite a few future ones, too!

**Comment by PepsRyuu** [comment]
Considering that Nollup is already a strange hybrid between Parcel and Webpack, and has different behaviour in how accept handlers resolve as we've discussed previously, I'm more inclined to pass the ids as an argument. Do you think the proposal above will work for you? Can you think of any other data that you would want passed into the accept callback?

Custom HMR runtime would require quite a lot of effort if it's to be compatible with the dev middleware. Don't know if I want to go that far just yet!

**Comment by rixo** [comment]
The idea with the custom runtime would be to ride over the existing communication protocol mainly (I confess next thing to know I'm gonna ask to emit build errors to the client, but let's do things in order ^^). My intuition is the effort on Nollup side should be minimal, but I'm not on my computer to really check and this discussion can wait...

I think the module ids solution would take us 95% there, probably all the most important cases. But I'm a perfectionist and I like 100%... If 2 files are changed at once ("save all" of IDE), will I be able to tell that both files have been hit by a direct update (thus allowing the shortcut for both) with this? Genuine question, I'm not sure I get all of your proposal.

I don't know of other such data that would be needed. And it took me a long time to figure this one from usage. So I think it's better to wait and see. More may come though, so I'd keep the object argument personally.

**Comment by PepsRyuu** [comment]
Prototype: https://github.com/PepsRyuu/nollup/pull/82

```disposed``` property given into the ```accept``` handler will provide a list of all modules on that branch that have been disposed. Think this will work for your use case?

**Comment by rixo** [comment]
Yes, I think. Testing this and the sourcemaps branch now.

**Comment by rixo** [comment]
Seems to work like a charm. Completely fulfills my need regarding this issue.

I've hit another issue though, with updated deps not being rebound correctly. Same with this version, and currently published version... Curious, I tried with my old test case project, and there deps get correctly rebound. I'm trying to pinpoint a minimal repro and open a new issue if needed.

**Comment by PepsRyuu** [comment]
Great! I'll look into publishing when I get a chance. :)

Keep me posted on the other issue.

**Comment by rixo** [comment]
The bug was mine. All good here!

**Comment by PepsRyuu** [comment]
Merged, release to follow after other small features merged.

**Comment by PepsRyuu** [comment]
Released in ```0.12.0```
